File: Javascript/234.回文链表.js
/*
 * @lc app=leetcode.cn id=234 lang=javascript
 *
 * [234] 回文链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if (!head) return true;
    let reverse = (pre, cur) => {
        if (!cur) return pre;
        const next = cur.next;
        cur.next = pre;
        pre = cur;
        cur = next;
        return reverse(pre, cur);
    }
    let fast = head.next;
    let slow = head;
    while(fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
    }

    let next = slow.next;
    slow.next = null;

    let newStart = reverse(null, next);

    for(let p = head, newP = newStart; newP != null; p = p.next, newP = newP.next) {
        if (p.val != newP.val) return false;
    }

    return true;
};
// @lc code=end

File: Go/LinkedList/234-回文链表.go
package LinkedList

// 快慢指针找到中间节点，反转后半部分，一次比对
func isPalindrome(head *ListNode) bool {
	// 找中间节点
	slow, fast := head, head
	// 添加一个额外头部节点，用于记录mid的前一个节点
	slowPrev := &ListNode{
		Val:  0,
		Next: head,
	}
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
		slowPrev = slowPrev.Next
	}
	// 前半部分链表最后一个节点的next置空
	slowPrev.Next = nil
	// 反转
	var prev *ListNode
	prev = nil
	tmp := slow
	for slow != nil {
		tmp = slow.Next
		slow.Next = prev
		prev = slow
		slow = tmp
	}
	// 遍历比对, len(head) <= len(prev)
	// prev最多只会比head多一个节点(奇数个节点时)，只对比len(head)范围内的节点就ok
	for head != nil {
		if head.Val != prev.Val {
			return false
		}
		head = head.Next
		prev = prev.Next
	}
	return true
}
